code/io/joy-sdl.cpp

```diff
@@ -145,6 +145,164 @@ namespace
 			}
 		}
 	}
+
+	void joy_set_button_state(int button, int state)
+	{
+		if (!Joy_inited)
+			return;
+
+		if (state != joy_buttons[button].actual_state) {
+			// Button position physically changed.
+			joy_buttons[button].actual_state = state;
+
+			if ( state )    {
+				// went from up to down
+				joy_buttons[button].down_count++;
+				joy_buttons[button].down_time = 0;
+
+				joy_buttons[button].state = 1;
+			} else {
+				// went from down to up
+				if ( joy_buttons[button].state )
+					joy_buttons[button].up_count++;
+
+				joy_buttons[button].state = 0;
+			}
+		} else {
+			// Didn't move... increment time down if down.
+			if (joy_buttons[button].state)
+				joy_buttons[button].down_time += JOY_POLLRATE;
+		}
+	}
+
+	void joy_set_hat_state(int position)
+	{
+		int state = 0;
+
+		if (!Joy_inited)
+			return;
+
+		for (int i=JOY_NUM_BUTTONS-1; i<JOY_TOTAL_BUTTONS; i++) {
+			switch (i) {
+				case JOY_HATBACK:
+					state = (position == SDL_HAT_DOWN) ? 1 : 0;
+					break;
+				case JOY_HATFORWARD:
+					state = (position == SDL_HAT_UP) ? 1 : 0;
+					break;
+				case JOY_HATLEFT:
+					state = (position == SDL_HAT_LEFT) ? 1 : 0;
+					break;
+				case JOY_HATRIGHT:
+					state = (position == SDL_HAT_RIGHT) ? 1 : 0;
+					break;
+				default:
+					break;
+			}
+
+			if (state != joy_buttons[i].actual_state) {
```

Comment:
This seems like a lot of duplicated code; could these two functions be merged, or maybe have this one call the other?